### PR TITLE
feat(native): implement R402 list-all-used-variables graph rule

### PR DESCRIPTION
**Pull request by djdanielsson**
## Summary

Split from #347 (per review). **PR 2 of 3** — R402 only.

> **Merge order:** merge #354 first, then rebase this PR onto `main`.

- Implements disabled-by-default play-scoped audit rule `R402`
- Traversal follows `CONTAINS` + `INCLUDE`/`IMPORT` only (not `DATA_FLOW`/`NOTIFY`) to avoid cross-play leakage
- Includes rule docs, catalog entry, shared audit infra, and focused tests

Depends on #354

## Review feedback addressed

- Play-scoped `no_log` for shared `include_tasks` (cross-play and dual-include same play)
- Complete Jinja ref extraction (`re.DOTALL`, filter args, multiline, parenthesized dotted paths)
- Path-aware sensitive redaction for dotted/bracket Jinja and bare `when`/`changed_when` expressions
- `play_scope` computed once per R402 evaluation; passed through `no_log_true_in_scope()`
- R402 doc example heading aligned with informational rule behavior
- RUF015 / RUF002 nitpicks in tests

## Known limitations (follow-up)

- Provenance resolver still uses CONTAINS-only scope chains; play vars in included files may report `source: external`
- L110 shared-include `no_log` without play context remains a cross-play edge case

## Test plan

- [x] `tox -e lint`
- [x] `tox -e unit -- tests/test_remaining_graph_rules.py -k R402 tests/test_graph_scanner.py`

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

- **New Features**
  - Added the R402 play-level audit rule to report used variables, with sensitive and `no_log` values redacted.
  - Added clearer handling and reporting for opted-in rules, including missing rule IDs.

- **Bug Fixes**
  - Improved credential and private-key redaction in URLs and metadata.
  - Added warnings for malformed audit metadata while preserving scan responses.
  - Improved rescanning of included tasks and enclosing plays.

- **Documentation**
  - Updated rule catalogs, remediation guidance, and development documentation.

- **Chores**
  - Removed the deprecated UI workflow validation environment.

### CODE REVIEW

**Comment by djdanielsson** [comment]
@coderabbitai full review

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/daemon/engine_server.py [hunk 1]
```diff
@@ -1547,7 +1553,16 @@ async def _session_graph_remediate(  # type: ignore[explicit-any]  # noqa: PLR09
...
+        opt_in_rules = graph_rule_opt_in_from_rule_configs(rule_configs)
+        rules, missing_opt_in = load_graph_rules(
+            rules_dir=native_rules_dir(),
+            opt_in_rule_ids=opt_in_rules,
+        )
+        if missing_opt_in:
+            logger.warning(
+                "Remediation: requested opt-in graph rules failed to load: %s",
+                ", ".join(missing_opt_in),
+            )
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/engine_server.py:2144]: Surfaced missing opt-in rule IDs as a graph-tier1 WARNING ProgressUpdate streamed to FixSession callers (matching the idempotency warning pattern), in addition to the existing server-side log. Fixed in e4eefa9f.

File: src/apme_engine/graph/sensitivity.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    r"password\s*[=:]|"
+    r"passwd\s*[=:]|"
+    r"api[_-]?key\s*[=:]|"
+    r"bearer\s+\S+|"
+    r"BEGIN\s+(?:RSA\s+)?PRIVATE\s+KEY|"
+    r"://[^:]+:[^@]+@|"
+    r"\$ANSIBLE_VAULT;|"
+    r"!vault\b"
+    r")"
+)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/sensitivity.py:42]: Broadened `_SCALAR_SECRET_RE` to match common PEM private-key headers (`OPENSSH`, `EC`, `ENCRYPTED`, etc.) via `(?:[A-Z]+\s+)?PRIVATE\s+KEY`. Added regression coverage in `test_value_looks_sensitive_openssh_private_key`. Fixed in e4eefa9f.

File: src/apme/v1/reporting_pb2.py [hunk 3]
```diff
@@ -2,7 +2,7 @@
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
 # NO CHECKED-IN PROTOBUF GENCODE
 # source: apme/v1/reporting.proto
-# Protobuf Python Version: 7.35.0
+# Protobuf Python Version: 6.33.5
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/reporting_pb2.py]: `reporting_pb2.py` is already generated with protobuf 7.35.1 (see file header and `ValidateProtobufRuntimeVersion` stamp), consistent with all other checked-in `*_pb2.py` stubs and `protobuf>=7.35.1,<8` in pyproject. No regeneration needed.

File: tests/test_generate_rule_catalog.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+        def fake_import(
+            name: str,
+            globals: Mapping[str, object] | None = None,
+            locals: Mapping[str, object] | None = None,
+            fromlist: Sequence[str] = (),
+            level: int = 0,
+        ) -> object:
+            if name == "yaml":
+                raise ImportError("PyYAML not installed")
+            return real_import(name, globals, locals, fromlist, level)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_generate_rule_catalog.py:257]: Added targeted `# noqa: A002` on the `globals`/`locals` parameters in `fake_import` so Ruff passes while keeping `__import__`-compatible names. Fixed in e4eefa9f.

File: src/apme_engine/graph/scanner.py [hunk 5]
```diff
@@ -320,6 +389,55 @@ def rescan_dirty(
...
+    for node_id in dirty_node_ids:
+        node = graph.get_node(node_id)
+        if node is None:
+            continue
+        if node.node_type in aggregate_node_types:
+            expanded.add(node_id)
+        for ancestor in graph.ancestors(node_id):
+            if ancestor.node_type in aggregate_node_types:
+                expanded.add(ancestor.node_id)
+    return frozenset(expanded)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/scanner.py:451]: Added `ContentGraph.positional_ancestor_ids()` (CONTAINS/INCLUDE/IMPORT reverse walk) and switched `expand_dirty_node_ids()` to use it so dirty included/imported tasks expand to enclosing PLAY nodes for play-scoped rules. Regression test in `test_expand_dirty_node_ids_includes_play_via_include_edge`. Fixed in e4eefa9f.

File: src/apme_engine/graph/sensitivity.py [hunk 6]
```diff
@@ -50,7 +50,7 @@
     r")"
 )
 _WORD_BOUNDARY_RE = re.compile(r"(?:^|[_.'\"\[])({})(?:[_.'\"\[\]]|$)".format("|".join(_SENSITIVE_WORDS)))
-_URL_USERINFO_RE = re.compile(r"^([^:]+)://([^/@]+):([^/@]+)@")
+_URL_USERINFO_RE = re.compile(r"^([^:/]+)://([^/]*)@")
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/sensitivity.py]: Restricted `_URL_USERINFO_RE` authority matching to exclude `?` and `#` (`^([^:/?#]+)://([^/?#]*)@`). Added regression tests preserving `@` in query and fragment values. Fixed in e4eefa9f.

File: src/apme_engine/graph/variable_helpers.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+    Returns:
+        Set of root variable names referenced in the Jinja expressions.
+    """
+    refs: set[str] = set()
+    for text in texts:
+        for m in _JINJA_VAR_RE.findall(text):
+            cleaned = m.strip().split("|")[0].split(".")[0].split("[")[0].strip()
+            if cleaned and cleaned.isidentifier():
+                refs.add(cleaned)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/variable_helpers.py]: Compound `{{ ... }}` bodies now fall back to `extract_bare_refs()` when the expression is not a single identifier, so `primary_host ~ : ~ db_port` yields both roots. Added `test_compound_jinja_expression_refs`. Fixed in e4eefa9f.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/rules/R402_list_all_used_variables_graph.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            return (desc_node.file_path, desc_node.line_start, desc_id)
+
+        for desc_id in sorted(task_desc_ids, key=_task_sort_key):
+            desc = graph.get_node(desc_id)
+            if desc is None:
+                continue
+            task_ids.append(desc_id)
+            templates, bare = collect_strings(desc)
+            task_refs = extract_jinja_refs(templates) | extract_bare_refs(bare)
+            refs_by_task[desc_id] = task_refs
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/rules/R402_list_all_used_variables_graph.py]: Split pre/post-task reference collection via `collect_timed_strings()` and classify the task’s own `register` in `changed_when`/`failed_when` as `runtime`, matching L039. Added `test_changed_when_register_source_is_runtime`. Fixed in 0c66043c.

File: src/apme_engine/graph/rules/R402_list_all_used_variables.md [hunk 9]
```diff
@@ -2,31 +2,45 @@
...
+usage data for audit and documentation purposes.
 
-**Planned** — no `_graph.py` implementation yet. This rule requires
-`VariableProvenanceResolver` to enumerate all variable references across the
-full task sequence and report them as an informational finding. It does not
-flag violations — it reports data for audit purposes.
+Traversal follows only structural ``CONTAINS`` edges plus dynamic
+``include_tasks`` / ``import_tasks`` links. It does **not** follow
+``DATA_FLOW`` or ``NOTIFY`` edges, so tasks in other plays that consume a
+registered fact are excluded from the play-level report.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/rules/R402_list_all_used_variables.md]: Added `ContentGraph.positional_ancestors()` and switched `no_log_true_in_scope()` to walk CONTAINS/INCLUDE/IMPORT ancestry so play-level `no_log` applies to included tasks. Documented in R402 rule docs and covered by `test_no_log_redacts_included_task_refs`. Fixed in 0c66043c.

File: tests/test_rule_catalog.py [hunk 10]
```diff
@@ -72,13 +72,16 @@ def test_collect_all_rules_non_empty_and_fields() -> None:
...
+
     rules = collect_all_rules()
     by_id = {r.rule_id: r for r in rules}
-    for rule_id in ("R402", "R404"):
-        assert rule_id in by_id, f"{rule_id} must be registered in collect_all_rules()"
+    for rule_id in sorted(DISABLED_BY_DEFAULT_GRAPH_RULE_IDS & by_id.keys()):
         assert by_id[rule_id].enabled is False, f"{rule_id} must remain disabled by default in catalog"
         assert by_id[rule_id].source == "native"
+    for rule_id in ("R402", "R404"):
+        assert rule_id in by_id, f"{rule_id} must be registered in collect_all_rules()"
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_rule_catalog.py]: Asserted R402/R404 are in `DISABLED_BY_DEFAULT_GRAPH_RULE_IDS`, that the full disabled set is registered in `collect_all_rules()`, and iterate over the complete disabled set when checking `enabled=False`. Fixed in 0c66043c.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/rules/R402_list_all_used_variables.md [hunk 11]
```diff
@@ -2,31 +2,47 @@
...
+  connection: local
+  vars:
+    http_port: 8080
+  tasks:
+    - name: Show port
+      ansible.builtin.debug:
+        msg: "Port is {{ http_port }}"
+```
+
+Disabled by default. Opt in via ``rule_id_list`` when loading graph rules.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/rules/R402_list_all_used_variables.md]: Renamed the example heading to "Example: report produced" so it matches R402's informational (non-violation) behavior. Fixed in e623ce9c.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/content_graph.py [hunk 12]
```diff
@@ -1424,6 +1424,80 @@ def ancestors(self, node_id: str) -> list[ContentNode]:
...
+            }
+        )
+        while True:
+            parents = sorted(
+                src for src, _, data in self.g.in_edges(current, data=True) if data.get("edge_type") in positional
+            )
+            if not parents:
+                break
+            parent_id = parents[0]
+            if parent_id in visited:
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/content_graph.py:1503]: Added `ContentGraph.play_scoped_node_ids()` and optional `play_context_id` to `positional_ancestors()` / `no_log_true_in_scope()`. R402 now passes the audited play id so shared include_tasks files follow the correct include path when plays have conflicting `no_log` values. Covered by `test_no_log_respects_play_context_for_shared_include`. Fixed in 57461181.

File: src/apme_engine/graph/variable_helpers.py [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+
+from __future__ import annotations
+
+import re
+
+from apme_engine.graph.content_graph import ContentGraph, NodeType
+
+TASK_TYPES: frozenset[NodeType] = frozenset({NodeType.TASK, NodeType.HANDLER})
+
+_JINJA_VAR_RE = re.compile(r"\{\{(.*?)\}\}")
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/variable_helpers.py]: Updated `_JINJA_VAR_RE` to compile with `re.DOTALL` and route every complete `{{ ... }}` body through `extract_bare_refs()` so filter arguments like `default(fallback)` and multiline expressions are captured (also benefits L039). Added `test_jinja_default_filter_argument_refs` and `test_multiline_jinja_expression_refs`. Fixed in 57461181.

**Comment by djdanielsson** [comment]
@coderabbitai full review

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/content_graph.py [hunk 14]
```diff
@@ -1424,6 +1424,122 @@ def ancestors(self, node_id: str) -> list[ContentNode]:
...
+            play_context_id: Optional play node id used to select the include
+                path for shared task files.
+
+        Returns:
+            Ancestor ``ContentNode`` instances from immediate parent toward root.
+        """
+        result: list[ContentNode] = []
+        visited: set[str] = set()
+        current = node_id
+        play_scope = self.play_scoped_node_ids(play_context_id) if play_context_id is not None else None
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/content_graph.py]: R402 now computes `play_scope = graph.play_scoped_node_ids(node_id)` once per `process()` call and passes it through `no_log_true_in_scope(..., play_scope=play_scope)`, avoiding repeated structural traversals for each task. Fixed in e87dd28a.

File: src/apme_engine/graph/content_graph.py [hunk 15]
```diff
@@ -1424,6 +1424,122 @@ def ancestors(self, node_id: str) -> list[ContentNode]:
...
+            parents = sorted(
+                src for src, _, data in self.g.in_edges(current, data=True) if data.get("edge_type") in positional
+            )
+            if play_scope is not None:
+                scoped_parents = [parent for parent in parents if parent in play_scope]
+                if scoped_parents:
+                    parents = scoped_parents
+            if not parents:
+                break
+            parent_id = parents[0]
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/content_graph.py:1502]: Replaced single-chain parent selection with `_no_log_any_play_scoped_path()`, which returns true when any in-scope include path inherits `no_log: true`. Shared tasks included twice in one play with conflicting `no_log` values are now redacted. Covered by `test_no_log_dual_include_same_play_redacts_shared_task`. Fixed in e87dd28a.

File: tests/test_remaining_graph_rules.py [hunk 16]
```diff
@@ -1,12 +1,13 @@
 """Unit and integration tests for Phase 2J+K GraphRules.
 
-Covers L056, R401, collection metadata rules (L087, L088, L096, L103–L105),
-and plugin/schema rules (L089, L090, L095).
+Covers L056, R401, R402, collection metadata rules (L087, L088,
+L096, L103–L105), and plugin/schema rules (L089, L090, L095).
 """
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_remaining_graph_rules.py:5]: Replaced the EN DASH in the module docstring range with a hyphen-minus (`L103-L105`) so Ruff RUF002 passes. Fixed in e87dd28a.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/variable_helpers.py [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+
+    Returns:
+        True when any in-scope ancestor path resolves ``no_log`` to true.
+    """
+    seen_paths: set[tuple[str, ...]] = set()
+
+    def walk(current_id: str, path: tuple[str, ...]) -> bool:
+        if current_id in path:
+            return False
+        next_path = path + (current_id,)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/variable_helpers.py]: Replaced tuple concatenation with unpacking (`next_path = (*path, current_id)`) to satisfy Ruff RUF005. Fixed in 94825bf.

**Comment by djdanielsson** [comment]
@coderabbitai full review
